software/firmware/europi.py

```diff
@@ -1,249 +1,278 @@
-        
+            if (ticks_ms() - self.last_pressed) > self.debounce_delay:
+                self.last_pressed = ticks_ms()
+                func()
+        # Both the digital input and buttons are normally high, and 'pulled'
+        # low when on, so here we use IRQ_FALLING to detect rising edge.
+        self.pin.irq(trigger=Pin.IRQ_FALLING, handler=bounce_wrapper)
+
     def reset_handler(self):
         self.pin.irq(trigger=Pin.IRQ_FALLING)
 
 
+class DigitalInput(DigitalReader):
+    def __init__(self, pin, debounce_delay=0):
+        super().__init__(pin, debounce_delay)
 
 
-#Define all the I/O using the appropriate class and with the pins used
-oled = Display(0,1,OLED_WIDTH,OLED_HEIGHT)
+class Button(DigitalReader):
+    def __init__(self, pin, debounce_delay=200):
+        super().__init__(pin, debounce_delay)
 
-k1 = Knob(27)
-k2 = Knob(28)
 
-b1 = DigitalInput(4)
-b2 = DigitalInput(5)
+class Display(SSD1306_I2C):
+    def __init__(self, sda, scl, width=OLED_WIDTH, height=OLED_HEIGHT, channel=I2C_CHANNEL, freq=I2C_FREQUENCY):
+        i2c = I2C(channel, sda=Pin(sda), scl=Pin(scl), freq=freq)
+        self.width = width
+        self.height = height
+
+        if len(i2c.scan()) == 0:
+            raise Exception(
+                "EuroPi Hardware Error:\nMake sure the OLED display is connected correctly")
+
+        super().__init__(self.width, self.height, i2c)
+
+    def clear(self):
+        """Clear the display upon call."""
+        self.fill(0)
+        self.show()
+
+    def centre_text(self, text):
+        """Split the provided text across 3 lines of display."""
+        self.fill(0)
+        # Default font is 8x8 pixel monospaced font which can be split to a
+        # maximum of 4 lines on a 128x32 display.
+        lines = str(text).split('\n')
+        maximum_lines = round(self.height / CHAR_HEIGHT)
+        if len(lines) > maximum_lines:
+            raise Exception(
+                "Provided text exceeds available space on oled display.")
+
+        padding_top = (self.height - (len(lines) * 9)) / 2
+        for index, content in enumerate(lines):
+            x_offset = int((self.width - ((len(content) + 1) * 7)) / 2) - 1
+            y_offset = int((index * 9) + padding_top) - 1
+            self.text(content, x_offset, y_offset)
+        oled.show()
+
+
+class Output:
+    def __init__(self, pin):
+        self.pin = PWM(Pin(pin))
+        # Set freq to 1kHz as the default is too low and creates audible PWM 'hum'.
+        self.pin.freq(1_000_000)
+        self._duty = 0
+
+    def _set_duty(self, cycle):
+        cycle = int(cycle)
+        self.pin.duty_u16(clamp(cycle, 0, MAX_UINT16))
+        self._duty = cycle
+
+    def voltage(self, voltage=None):
+        """Set the output voltage to the provided value within the range of 0 to 10."""
+        if voltage is None:
+            return self._duty / MAX_UINT16
+        if 0 > voltage or voltage > 10:
+            raise ValueError(

```

Comment:
Do we actually need to `raise` for this?

Reply:
Perhaps clamping at 10 would be better. I think the original intention was to bail if voltage was > 10 with the assumption that it's a user script logic error that should be fixed. Gracefully clamping is better that a killing the script.